jvm/benchmarks/Main.hs

```diff
@@ -49,6 +50,13 @@ incrHaskell x = return (x + 1)
 
 foreign import ccall unsafe getpid :: IO Int
 
+-- | Execute the given IO in a separate thread, then wait for it to finish
+forkWait :: IO a -> IO ()
+forkWait io = do
+  handle <- newEmptyMVar
+  _ <- forkFinally (runInBoundThread io) (\_ -> putMVar handle ())
+  takeMVar handle >>= return
```

Comment:
I'm not entirely sure what the reason for spawning a thread then immediately waiting for it would be. I'd imagine this has exactly the same semantics as just doing `io`. But probably I'm missing something.

At any rate, you shouldn't implement this sort of logic yourself. The [async](https://www.stackage.org/lts-16.5/package/async) implements high level primitives which do the `MVar` thing, and error handling, and such. You should reach for these.

Reply:
> At any rate, you shouldn't implement this sort of logic yourself. The async implements high level primitives which do the MVar thing, and error handling, and such. You should reach for these.

That's good advice in general. In this particular case, I think it is better to not spawn any threads. The need to spawn the thread comes from trying to measure how much time attaching and detaching the thread takes. The main thread is already attached to the JVM so we can't measure it there.

I think the best solution is to have the benchmark setup detach the main thread before running the benchmark.